org.openhab.binding.zigbee/src/main/java/org/openhab/binding/zigbee/discovery/internal/ZigBeeDefaultDiscoveryParticipant.java

```diff
@@ -0,0 +1,66 @@
+package org.openhab.binding.zigbee.discovery.internal;
+
+import java.util.Map;
+import java.util.Set;
+
+import org.eclipse.smarthome.config.discovery.DiscoveryResult;
+import org.eclipse.smarthome.config.discovery.DiscoveryResultBuilder;
+import org.eclipse.smarthome.core.thing.Thing;
+import org.eclipse.smarthome.core.thing.ThingTypeUID;
+import org.eclipse.smarthome.core.thing.ThingUID;
+import org.openhab.binding.zigbee.ZigBeeBindingConstants;
+import org.openhab.binding.zigbee.discovery.ZigBeeDiscoveryParticipant;
+import org.openhab.binding.zigbee.discovery.ZigBeeNodePropertyDiscoverer;
+import org.openhab.binding.zigbee.internal.ZigBeeThingTypeMatcher;
+import org.osgi.service.component.annotations.Component;
+
+import com.zsmartsystems.zigbee.ZigBeeNode;
+
+/**
+ * The default ZigBee discovery participant
+ *
+ * @author Chris Jackson
+ *
+ */
+@Component(immediate = true)
+public class ZigBeeDefaultDiscoveryParticipant implements ZigBeeDiscoveryParticipant {
```

Comment:
I think that it is useful to provide the `ZigBeeNode` object to `ZigBeeDiscoveryParticipant`s; still, when imagining to implement such a participant, I think that it would probably look like this class here, just with a different variant of the `ZigBeeThingTypeMatcher` plugged in.

I therefore think that it would be useful to extract an interface from the `ZigBeeThingTypeMatcher`, and make this class parametric in the thing type matcher, so that other 'participants' could reuse the remaining code that is implemented in this class easily. (One could, e.g., introduce an `AbstractZigBeeDiscoveryParticipant` where the thing type matcher can be set to one's own implementation of the thing type matching interface.)

What do you think?

Reply:
> I think that it is useful to provide the ZigBeeNode object to ZigBeeDiscoveryParticipants; still, when imagining to implement such a participant

I agree, but where exactly do you mean? The ```ZigBeeNode``` is already provided to the methods of this interface, so I don't understand what is missing.

Reply:
What I meant is that some implementations of `ZigBeeDiscoveryParticipant` would also use a `ZigBeeNodePropertyDiscoverer` to extract node properties, and would also use the same logic in the `createResult` method as in this class.

So this logic could be extracted into a common abstract super class. But this could also be added separately after this PR has been merged in another PR.

Reply:
> What I meant is that some implementations of ZigBeeDiscoveryParticipant would also use a ZigBeeNodePropertyDiscoverer to extract node properties

I don't think so - all services are first discovered by the property discoverer - these properties are passed to the discovery participant, so it already has everything. It also has the ZigBeeNode, so it is able to read any attribute it likes. It should not need to discover the same properties again?

Reply:
> I don't think so - all services are first discovered by the property discoverer - these properties are passed to the discovery participant, so it already has everything. It also has the ZigBeeNode, so it is able to read any attribute it likes. It should not need to discover the same properties again?

This is the case for the `createResult` method, but not for the `getThingTypeUID` method; `getThingTypeUID` only gets the node, and has to use a `ZigBeeNodePropertyDiscoverer` to get the properties (at least that is how it is done here).

Maybe the solution would be to add the properties also to the signature of `getThingTypeUID` to avoid this?

Reply:
> In fact the properties are not required in the participant at all. The participant can simply request them itself directly using the clusters. Normally, it will just require the data that has already been requested by the property discoverer, but remember that these are cached by the framework, so we don’t need to read them again. (by this, I mean they will not be read again over the air since the next call to the cluster will simply return the cached value)

Right. Still, as it is currently implemented, a `ZigBeeNodePropertyDiscoverer` is constructed and used twice - once in the `ZigBeeDiscoveryService`, and once in the `ZigBeeDefaultDiscoveryParticipant`, as the `getThingTypeUID` does not get the properties passed.

Reply:
> So you want me to remove it from the participant? I’m not sure it’s necessary, but if that’s what you want, I’ll do this...

No, there we misunderstood each other.

What I originally suggested was to make the code of the `ZigBeeDefaultDiscoveryParticipant` reusable by other discovery participants :) From there, we somehow digressed to discussing the signatures of the methods in the `ZigBeeDiscoveryParticipant` interface.

Let me flesh out my suggestion with concrete code; I will add a comment with this here when I am done.

Reply:
Here's my suggestion - based on the assumption that most of the logic that is currently in the `ZigBeeDefaultDiscoveryParticipant` will be used by other discovery participants as well - just the thing type matching will differ.

One would have an `AbstractZigBeeDiscoveryParticipant` as follows (this is the `ZigBeeDefaultDiscoveryParticipant` made parametric in a thing type matcher):

```java
protected abstract class AbstractZigBeeDiscoveryParticipant implements ZigBeeDiscoveryParticipant {

    private final ThingTypeMatcher thingTypeMatcher;

    public AbstractZigBeeDiscoveryParticipant(ThingTypeMatcher thingTypeMatcher) {
        this.thingTypeMatcher = thingTypeMatcher;
    }

    @Override
    public Set<ThingTypeUID> getSupportedThingTypeUIDs() {
        return thingTypeMatcher.getSupportedThingTypeUIDs();
    }

    @Override
    public ThingTypeUID getThingTypeUID(ZigBeeNode node) {
        ZigBeeNodePropertyDiscoverer propertyDiscoverer = new ZigBeeNodePropertyDiscoverer();
        Map<String, String> properties = propertyDiscoverer.getProperties(node);

        return thingTypeMatcher.matchThingType(properties);
    }

    @Override
    public DiscoveryResult createResult(ThingUID bridgeUID, ZigBeeNode node, Map<String, Object> properties) {
        ThingUID thingUID = new ThingUID(ZigBeeBindingConstants.BINDING_ID, bridgeUID,
                node.getIeeeAddress().toString().toLowerCase().replaceAll("[^a-z0-9_/]", ""));

        String label;
        // If we know the manufacturer and model, then give this device a name and a thing type
        if ((properties.get(Thing.PROPERTY_VENDOR) != null) && (properties.get(Thing.PROPERTY_MODEL_ID) != null)) {
            label = properties.get(Thing.PROPERTY_VENDOR) + " " + properties.get(Thing.PROPERTY_MODEL_ID);
        } else {
            label = "Unknown ZigBee Device";
        }

        ThingTypeUID thingTypeUID = getThingTypeUID(node);
        if (thingTypeUID == null) {
            return null;
        }

        return DiscoveryResultBuilder.create(thingUID).withThingType(thingTypeUID).withProperties(properties)
                .withBridge(bridgeUID).withLabel(label).build();
    }

}
```

The `ZigBeeDefaultDiscoveryParticipant` would then simply look like this:

```java
@Component(immediate = true)
public class ZigBeeDefaultDiscoveryParticipant extends AbstractZigBeeDiscoveryParticipant {

    public public ZigBeeDefaultDiscoveryParticipant() {
        super(new ZigBeeThingTypeMatcher());
    }

}
```

Reply:
In my above suggestion, `ThingTypeMatcher` would be an interface extracted from the `ZigBeeThingTypeMatcher`:

```java
public interface ThingTypeMatcher {
  ThingTypeUID matchThingType(Map<String, String> properties);
  Set<ThingTypeUID> getSupportedThingTypeUIDs();
}
```

Reply:
I am surprised that you intend to reuse most of the functionality from the default participant - I understood from previous discussions with DT that the plan was to do this differently. I don't mind if we do it this way.

I think it we are going to have many participants using the PropertyDiscovery, then this needs more work to avoid blocking the network, but let's leave this for another day.

Reply:
> I am surprised that you intend to reuse most of the functionality from the default participant - I understood from previous discussions with DT that the plan was to do this differently. I don't mind if we do it this way.

OK. As mentioned above, we can either include it in this PR, or I can add as addon PR later, both fine for me.

> I think it we are going to have many participants using the PropertyDiscovery, then this needs more work to avoid blocking the network, but let's leave this for another day.

Do I understand you correctly that you refer to the use of  the `ZigBeeNodePropertyDiscoverer` here? (If so, wouldn't a solution here be what we discussed above - to have the properties map not only passed to `createResult`, but also to `getThingTypeUID`? Because then the participants would not need to use the PropertyDiscovery.)

Reply:
Either way, it needs some more thought (one day). The main reason that we added the participant interface was that the properties gathered by the default implementation were considered to not cover all possibilities. So whatever we do, the assumption therefore must be that participants may be requesting other data. This also means that participants will take a long time to return (maybe 10s of seconds, depending on the device).

If there are a lot of participants, then this can quickly get quite large.  Note that this is why I add a default discovery result before calling the participants and then remove it again afterwards, as I consider this process may take a long time in some cases.

If we pass in the properties map, and the property a participant is looking for doesn't exist in the map, what does it assume? It can assume the device didn't report it, or it can assume the default implementation did not retrieve this property, and then ask again. Either way, unless we say that participants are not allowed to request this data, the result is the same. It will use the node/cluster to request the data - if the property data is already known, then it is returned immediately from the cluster cache (same as if we had passed in the properties) - if it's not known, then it is requested again from the device.